src/linker/Linker.Steps/DiscoverCustomOperatorsHandler.cs

```diff
@@ -130,7 +130,18 @@ TypeDefinition NonNullableType (TypeReference type)
 			// Unwrap Nullable<T>
 			Debug.Assert (typeDef.HasGenericParameters);
 			var nullableType = type as GenericInstanceType;
-			Debug.Assert (nullableType != null && nullableType.HasGenericArguments && nullableType.GenericArguments.Count == 1);
+			// The original type reference might be a TypeSpecification like array of Nullable<T>
+			// that we need to unwrap until we get to the Nullable<T>
+			while (nullableType == null) {
+				var typeSpec = type as TypeSpecification;
+				Debug.Assert (typeSpec != null);
+				if (typeSpec == null)
```

Comment:
This doesn't make sense. Either `null` is a valid value here and we should guard against it, or it's not and we should assert that it can never happen.

Reply:
The extra null check was there to keep the operator in case I'm wrong and there's a way for it to not be a TypeSpecification - but maybe it's better to let it throw a nullref so we'd catch the missing case.

Regarding your other feedback: there's an `IsTypeSpecification` check but it's not public: https://github.com/jbevain/cecil/blob/master/Mono.Cecil/TypeReference.cs#L322. I did change the GenericInstanceType type check to use `IsGenericInstanceType` instead.

Reply:
Thanks -- I'm still trying to understand how Cecil works. So if you call `TryResolve` on the type reference that holds an array you get back the underlying array element type?

Is it the same for pointers? What about function pointers?

Reply:
Yup, it gives back the array element type. It's pretty confusing and we already have our own helper for cases where this isn't the right behavior:

https://github.com/mono/linker/blob/fb9ef5b60a953842a597f0c71ffb862ecaae4af7/src/linker/Linker.Dataflow/MethodBodyScanner.cs#L921-L930

Pointers do the same (Foo* gives Foo), and function pointers have element type null.